Added support for NaNs.

grahambinns opened the pull request:
Add support for NaN when doing deep-equal on objects that may contain them. As far as JS is concerned NaN is never equal to anything else, including itself, so the following would not match when passed through deep-equal:

```
{a: NaN}
{a: NaN}
```

Even though they are to all intents and purposes the same.
To work around this I've added a call to Object.is() to check the identity of the actual and expected objects passed to deepEqual().

bathos commented:
I think the tests failed in node <=0.6 because Object.is didn't exist yet. I'd like to see this too though. Is the obj === obj check a reasonable fallback?

grahambinns commented:
Sorry for leaving this dangling for so long.

I've updated the check so that:

 - Where Object.is() can be used, it is used.
 - Where Object.is() isn't available, fall back to checking for numbers that don't equal themselves — NaN doesn't equal itself but typeof returns "number", so we can actually infer the value from that.

isiahmeadows commented:
For what it's worth, I still usually do `x !== x` for my `NaN` checks. It's shorter than Object.is, and it's impossible to fool in any browser.

iamjochem commented:
hi there, 

I just ran into a situation (using `tape`) where I need to compare objects that contain `NaN`s, which obviously didn't work (regardless of the value of the "strict" option). 

It would really neat to be able to do this check with `deep-equal` rather than having to resort to manually deconstructing the objects and performing `isNaN()` checks in the appropriate places as well making sure the rest of the object stucture/values are equal. 

Any chance of getting this one merged? (and used in a new version of `tape`)?

isiahmeadows commented:
Ping @substack?

GregBee2 commented:
I support NaN in deepequal should return true

ljharb commented:
@grahambinns please check the "allow edits" checkbox on the right hand side of this PR - I've rebased this and updated it to use the `object-is` package.

However, node's `assert.deepEqual(NaN, NaN)` throws - so that's the correct behavior - but `assert.deepStrictEqual` does not (also, `assert.deepEqual(0, -0)` does not throw, but `assert.deepStrictEqual(0, -0)` does).

grahambinns commented:
> @grahambinns please check the "allow edits" checkbox on the right hand side of this PR - I've rebased this and updated it to use the `object-is` package.

Done.

ljharb commented:
I've updated this with test cases that match what node does, for NaNs and `0`/`-0`.
